### ISSUE TITLE
Build fails consistently

### ISSUE THREAD

**Issue by IoTFier** [synthetichealth/synthea#1043]
### What happened?

I updated my master repo to the latest commit at master - 7876535 - Update HAPI and FHIR exporters.
But this time the build has been failing consistently every time I am trying to build the code.

![image](https://user-images.githubusercontent.com/26844611/162605326-d50c217c-8a4b-478b-a976-5eec2242effd.png)

### Environment

```markdown
- OS: Win 11 64 bit
- Java: java --version                               
openjdk 17.0.1 2021-10-19
OpenJDK Runtime Environment Temurin-17.0.1+12 (build 17.0.1+12)
OpenJDK 64-Bit Server VM Temurin-17.0.1+12 (build 17.0.1+12, mixed mode, sharing)
```

### Relevant log output

```shell
.\gradlew.bat build check test               
Starting a Gradle Daemon, 1 incompatible and 1 stopped Daemons could not be reused, use --status for details

> Task :test

org.mitre.synthea.export.CCDAExporterTest > testCCDAExport FAILED
    java.lang.AssertionError: Validation of exported CCDA failed: Illegal operation (self.realmCode->isEmpty())|Illegal operation (self.realmCode->isEmpty())|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS)) expected:<0> but was:<10>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:645)
        at org.mitre.synthea.export.CCDAExporterTest.testCCDAExport(CCDAExporterTest.java:62)

org.mitre.synthea.export.CCDAExporterTest > testExportWithNoPreferredWellnessProvider FAILED
    java.lang.AssertionError: expected:<0> but was:<1>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:645)
        at org.junit.Assert.assertEquals(Assert.java:631)
        at org.mitre.synthea.export.CCDAExporterTest.testExportWithNoPreferredWellnessProvider(CCDAExporterTest.java:92)

508 tests completed, 2 failed, 5 skipped

> Task :test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///{path}/synthea/build/reports/tests/test/index.html

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 4m 32s
15 actionable tasks: 2 executed, 13 up-to-date
```

**Comment by jawalonoski** [comment]
I do not have a Windows environment.

I have been unable to recreate this issue using Java 17.0.2. Still trying...

**Comment by jawalonoski** [comment]
~ 15000 patients generated without encountering any CCDA errors....

**Comment by IoTFier** [comment]
Great. What am I missing on my end?

**Comment by IoTFier** [comment]
Referring to - https://github.com/synthetichealth/synthea/issues/983#issuecomment-1013262333 - @eedrummer  was getting intermittent error like my experience until this commit; now I am totally stuck.

**Comment by jawalonoski** [comment]
I don't know. I'm going to try a few different things to see if I can get this to occur...

**Comment by eedrummer** [comment]
I'm not sure what is going on either. I thought that the CCDA export errors were largely addressed. What was previously causing issues was the situation where veterans would have their provider reset at age 20, which created a brief window where they had no wellness provider. 

Given that the no wellness provider test is failing in this case, there may be another way for individuals to have no wellness provider that the CCDA export logic can't handle.

**Comment by jawalonoski** [comment]
The test process uses random generation during export validation. Meaning, the test runs are randomized. Why? Because there are so many possible ways for errors to sneak in that exporting every single possible patient is impossible. So, we randomize the unit tests. I'm sure people hate this. And it does make debugging unit test failures difficult some times. But we find a lot of issues this way.

**Comment by IoTFier** [comment]
I appreciate the design intent. Totally makes sense.
The question is - can the issue(s) causing test failures be logged in the output.
Right now we all we see is test names. 
Thoughts?

**Comment by jawalonoski** [comment]
Well, in the test report (the screen shot you sent), often times if you click on the test failure there is additional output information in stdout that is not shown on the command line.

**Comment by eedrummer** [comment]
There is also a bit of a Catch 22 here. To tell why an export test is failing, we would need to... export the objects that are causing the failures. That said, it may make sense to run the JSON Exporter when one of the non-JSON Exports fail.

**Comment by jawalonoski** [comment]
Well, sometimes you can export it, the result just won't validate.

**Comment by IoTFier** [comment]
understood. what specific that we need to help in publishing the cause of the error while correcting or tuning any parameter to get the build completed successfully. Then in the next update, with the captured details can help in fine tuning either the code or test case or both as needed.
??

**Comment by jawalonoski** [comment]
I think @eedrummer is correct, we need to modify the unit tests to save the records that fail to validate (assuming they can even export).

**Comment by garyisaac** [comment]
Looks like this could be a memory issue. My machine ran out of memory while installing Ubuntu in VirtualBox. I restarted the device, installed Ubuntu, and successfully built Synthea. Decided to see if I could build it on Windows. I restarted my machine and tried building it the same way as when it failed. I successfully built Synthea in Windows after freeing up the memory.

**Comment by eedrummer** [comment]
An update for those who are interested in this issue:

I'm working on the feature I described earlier in the conversation, where a failure in the FHIR exporters or C-CDA export tests causes a bunch of information to be written out to disk. You can find that on the `export-on-failure` branch: https://github.com/synthetichealth/synthea/tree/export-on-failure

After doing some work there, I discovered that the version of MDHT that is packaged with Synthea behaves inconsistently. I had a case where the C-CDA test failed and the XML was written out to a file. I ran that file back through MDHT and it passed.

Further, while the documents produced by Synthea pass the test suite most of the time, they do not pass the C-CDA validator at: https://site.healthit.gov/sandbox-ccda/ccda-validator

I'm going to be working on updating our version of MDHT and the C-CDA templates to see if we can get something more consistent and valid.

**Comment by IoTFier** [comment]
Hi @eedrummer - your support and update are appreciated. Have a wonderful day ahead!

**Comment by IoTFier** [comment]
Hi @eedrummer, I updated my build to the latest commit "4b4dcf9 - Merge pull request #959 from synthetichealth/hiv"

.\gradlew.bat build check test                 pwsh  11:55:25 
Starting a Gradle Daemon, 1 incompatible and 1 stopped Daemons could not be reused, use --status for details

> Task :test

org.mitre.synthea.export.CCDAExporterTest > testCCDAExport FAILED
    java.lang.AssertionError: Validation of exported CCDA failed: Illegal operation (self.realmCode->isEmpty())|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS)) expected:<0> but was:<10>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:645)
        at org.mitre.synthea.export.CCDAExporterTest.testCCDAExport(CCDAExporterTest.java:62)

org.mitre.synthea.export.CCDAExporterTest > testExportWithNoPreferredWellnessProvider FAILED
    java.lang.AssertionError: expected:<0> but was:<1>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:645)
        at org.junit.Assert.assertEquals(Assert.java:631)
        at org.mitre.synthea.export.CCDAExporterTest.testExportWithNoPreferredWellnessProvider(CCDAExporterTest.java:92)

510 tests completed, 2 failed, 5 skipped

> Task :test FAILED

FAILURE: Build failed with an exception.

But, I am still getting error when doing the build -
![image](https://user-images.githubusercontent.com/26844611/167221678-581c217f-0988-4eeb-9e61-4d92b05a4bb5.png)
![image](https://user-images.githubusercontent.com/26844611/167221704-9a3a4a95-be14-4f70-aaf2-a5bfe3e861c8.png)
![image](https://user-images.githubusercontent.com/26844611/167221733-6135d646-7cde-4acd-8c22-55c546b48257.png)
![image](https://user-images.githubusercontent.com/26844611/167221789-f8a3b68d-7abe-4108-ac0e-90cf6623bcaa.png)
![image](https://user-images.githubusercontent.com/26844611/167221858-a3fc1677-b778-4c6a-bd1b-819a40823bc4.png)
![image](https://user-images.githubusercontent.com/26844611/167221897-e555d51d-20a8-47ed-a81a-f9c6d10fc882.png)

**Comment by IoTFier** [comment]
Hello @eedrummer and @jawalonoski ,

I could run build successfully after updating my code to the latest commit - dbd7df3 - Merge pull request #1028 from synthetichealth/metabolic_syndrome_refactor.

@eedrummer, is the fix for issues related to CCDA tests are addressed in this commit ?

Thank you very much.

**Comment by jawalonoski** [comment]
> @eedrummer, is the fix for issues related to CCDA tests are addressed in this commit ?

Not to speak for @eedrummer, but _"maybe"_ is probably the best answer.

We don't know why it was failing (see comments above and in the PR thread) because files that would fail validation during unit test would pass validation on subsequent attempts... So, we updated the CCDA library we are using (MDHT) and we now capture "bad" files if the issue happens again.

**Comment by eedrummer** [comment]
I think that @jawalonoski is right. We might of fixed the issue. This is a bit of a [Heisenbug](https://en.wikipedia.org/wiki/Heisenbug).

**Comment by IoTFier** [comment]
@jawalonoski , @eedrummer  - thank you for your updates. Really appreciate it.
Let's check the build when next two commits takes place on master. That should solidify your guess. 
Thoughts?

**Comment by jawalonoski** [comment]
That's the plan.

**Comment by IoTFier** [comment]
Build failed for commit - f9d6afa - Remove unique index on Resource.ResourceSurrogateId (#2626)

![image](https://user-images.githubusercontent.com/26844611/170883016-fb704db3-159f-4c46-8e4b-3658921a5ac6.png)

testCCDAExport
java.lang.AssertionError: Validation of exported CCDA failed: Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS))|Cannot find operation (excluding(OclVoid)) for the type (Set(TS)) expected:<0> but was:<10>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.mitre.synthea.export.CCDAExporterTest.testCCDAExport(CCDAExporterTest.java:64)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at jdk.internal.reflect.GeneratedMethodAccessor59.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jdk.proxy2/jdk.proxy2.$Proxy5.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$2.run(TestWorker.java:176)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
testExportWithNoPreferredWellnessProvider
java.lang.AssertionError: expected:<0> but was:<1>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at org.mitre.synthea.export.CCDAExporterTest.testExportWithNoPreferredWellnessProvider(CCDAExporterTest.java:94)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at jdk.internal.reflect.GeneratedMethodAccessor59.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jdk.proxy2/jdk.proxy2.$Proxy5.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$2.run(TestWorker.java:176)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)

**Comment by IoTFier** [comment]
Sorry about that - Commit info corrected above - 
The zip file below is the failed exports folder -
[failed_exports.zip](https://github.com/synthetichealth/synthea/files/8806278/failed_exports.zip)

Should I delete this folder everytime I do a new build? or will the build process take care emptying it before writing a new set of files there?

**Comment by eedrummer** [comment]
The folder is not automatically cleaned. Files are output using the person UUID, so they will be unique and not clobber each other between test suite runs.

**Comment by IoTFier** [comment]
Hello @jawalonoski, @eedrummer 

My build failed with following errors after updating master to the commit "1fc0868 - Merge pull request #1089 from synthetichealth/bfd-clm-grp-init-fix" - 
![image](https://user-images.githubusercontent.com/26844611/173803324-10d40277-f760-4f27-a000-fd7e6803d9e0.png)

Attached below is the zip of the _Failed Exports_ folder -
[failed_exports_06.14.2022.zip](https://github.com/synthetichealth/synthea/files/8908312/failed_exports_06.14.2022.zip)

**Comment by jawalonoski** [comment]
Do you have any local changes? I'm asking because you have other unit test failures that I don't normally see.

The C-CDA file that "failed" validation is valid according to https://site.healthit.gov/sandbox-ccda/ccda-validator and it is valid when I retest locally with MDHT.

**Comment by IoTFier** [comment]
I don't have any local changes besides synthea.properties and backups from failed exports folder.
My output to building ndjson output. (exporter.fhir.bulk_data = true)
Should I upload my synthea.properties file?

**Comment by jawalonoski** [comment]
No, you do not need to upload synthea.properites. I'm curious about the ndjson setting though, I wonder if that is related to the `PersonTest.testPersonFhir*Recreation` failures.

**Comment by jawalonoski** [comment]
Yes, the ndjson setting definitely breaks those `PersonTest.testPersonFhir*Recreation` unit tests and the `resolveAndExportEncounterCodes` test. I'll start working on a fix for those...

**Comment by IoTFier** [comment]
BUILD SUCCESSFUL in 6m 56s
16 actionable tasks: 11 executed, 5 up-to-date

No CCDA error.  I am not sure if ndjson has to do anything with CCDA failure.

In your ndjson fix, I was wondering if it will be possible to address #1069 ?
Thank you.

**Comment by IoTFier** [comment]
For commit - "acda325 - Merge pull request #1095 from synthetichealth/ndjson_unitest_fix" - I got successful builds for both with and without ndjson. Thank you @jawalonoski , @eedrummer 

Will update you with my findings once "b2233bb - Remove dangling comment."  is merged into Master.
Thanks

**Comment by IoTFier** [comment]
Build successful for commit - "3c5875e - Merge pull request #1099 from synthetichealth/rif-export-minimal-set" with ndjson.

**Comment by IoTFier** [comment]
Build successful for commit -"a522848 - Merge pull request #1106 from synthetichealth/rif_pde_costs" with ndjson.
Will really appreciate if the next master merge can include #1069

@jawalonoski, @eedrummer - 
Can we conclude the issue related to CCDA error is addressed now?

**Comment by IoTFier** [comment]
@eedrummer , @jawalonoski  -
Build failed for commit -"9fafd86 - Merge pull request #1108 from synthetichealth/notes_alcoholic_fix"
Failed exports folder > [failed_exports_commit_9fafd86.zip](https://github.com/synthetichealth/synthea/files/9109758/failed_exports_commit_9fafd86.zip)

![image](https://user-images.githubusercontent.com/26844611/178931097-f61e4143-19ca-4684-8aea-7f3c8ce660a4.png)

**Comment by jawalonoski** [comment]
> @eedrummer , @jawalonoski - Build failed for commit -"9fafd86 - Merge pull request #1108 from synthetichealth/notes_alcoholic_fix" Failed exports folder > [failed_exports_commit_9fafd86.zip](https://github.com/synthetichealth/synthea/files/9109758/failed_exports_commit_9fafd86.zip)

All of the files in `failed_exports` validate without errors.

**Comment by IoTFier** [comment]
Build successful with exporter.fhir.bulk_data = false 
![image](https://user-images.githubusercontent.com/26844611/178990251-82a37f5a-c774-47d2-bd28-73e04a74aded.png)

Looks like a bug in when building ndjson output. 
Any clue on why files are getting written to the failed_exports folder in that case?

**Comment by IoTFier** [comment]
Build successful for commit - "d4af945 - Merge pull request #1059 from synthetichealth/fixed-record-enhancements"
for both ndjson and json output

**Comment by IoTFier** [comment]
Hello @eedrummer , @jawalonoski -
Build failed for commit - "5d93079 - Merge pull request #1109 from synthetichealth/fixed-records-fix"
Failed exports folder >
[failed_exports - Aug 27 2022.zip](https://github.com/synthetichealth/synthea/files/9437593/failed_exports.-.Aug.27.2022.zip)

Synthea.Properties | exporter.fhir.bulk_data = true

![image](https://user-images.githubusercontent.com/26844611/187023022-ce0bcfc4-5027-4882-8e8e-f9829c33027d.png)
![image](https://user-images.githubusercontent.com/26844611/187022982-c66410f5-6d10-428d-8b0b-2f3060657783.png)

**Comment by IoTFier** [comment]
Hello @jawalonoski , @eedrummer 

Build failed for commit - "fb5add0 - Merge pull request #1132 from synthetichealth/csv_encoding"

Synthea.Properties | exporter.fhir.bulk_data = false

Failed exports folder > [failed_exports - Sept 1 2022.zip](https://github.com/synthetichealth/synthea/files/9469210/failed_exports.-.Sept.1.2022.zip)

![image](https://user-images.githubusercontent.com/26844611/187896952-cbde904c-7ac9-481a-9bd4-f2cd577f0dbf.png)
![image](https://user-images.githubusercontent.com/26844611/187897166-6f5736d9-cd67-45e6-9b7b-dfb671d4bdbd.png)

Can you please help with this issue? 

Thank you

**Comment by eedrummer** [comment]
@IoTFier different commits seem to be angering the MDHT library in different ways. I'm honestly kind of stumped as to why some unrelated commits break the CCDA tests. My guess at this point is maybe memory pressure or version of java, but it's kind of a shot in the dark.

What version of java are you using? 

Also you could try to tweak `test.maxHeapSize` in `build.gradle` to see if that helps.

**Comment by jawalonoski** [comment]
I don't understand it either. Every time I revalidate the offending CCDA documents with MDHT they successfully pass validation. Perhaps we should start executing the CCDA validation in a separate process using `Runtime` or something ridiculously overkill to isolate random MDHT failures from actual CCDA errors.

```java
Runtime r = Runtime.getRuntime();
Process p = r.exec("java ccdaValidationApp " + ccdaFilename);
```

**Comment by IoTFier** [comment]
@eedrummer I changed the memory to 8192m but didn't help. I don't think it is HeapSize related issue.
How do I start a clean (re)build again when running .\gradlew.bat build check test? 

<img width="1424" alt="image" src="https://user-images.githubusercontent.com/26844611/187958545-5afa52da-6b7d-4db0-8980-cbc2987fc508.png">
<img width="530" alt="image" src="https://user-images.githubusercontent.com/26844611/187958661-ac125dcc-fdc8-426a-8553-33abcee0c54f.png">

**Comment by IoTFier** [comment]
@jawalonoski, how do I move forward on my win PC  with the build failing against lastest commit in the master?

**Comment by jawalonoski** [comment]
I managed to get a Java 17 `master` build on Windows 10. No errors.

I also generated 100 patients using bulk data without errors.

Do you have any local changes?

```
git status
git diff
```

If so, undo your local changes (save them somewhere, maybe using `git stash` if you know how to do that) and see if it builds with a clean copy.

If you make changes for generating data (e.g., bulk data parameters) you don't necessarily have to rerun the unit tests first.

**Comment by IoTFier** [comment]
The only local change that I have is in the Synthea.properties where - 

- exporter.ccda.export = true
- exporter.fhir.bulk_data = true
- exporter.text.export = true
- exporter.text.per_encounter_export = true
- exporter.clinical_note.export = true

and failed_exports folder has 3 json files which caused error last night.

**_Output of git stauts_**
 git status                                          pwsh  10:12:20 
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   src/main/resources/synthea.properties

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        failed_exports/

no changes added to commit (use "git add" and/or "git commit -a")

**_Output of git diff_**
git diff                                           pwsh  10:12:22 
diff --git a/src/main/resources/synthea.properties b/src/main/resources/synthea.properties
index 1a15bee5a..de94a2ae7 100644
--- a/src/main/resources/synthea.properties
+++ b/src/main/resources/synthea.properties
@@ -10,14 +10,14 @@ exporter.years_of_history = 10
 exporter.split_records = false
 exporter.split_records.duplicate_data = false
 exporter.metadata.export = true
-exporter.ccda.export = false
+exporter.ccda.export = true
 exporter.fhir.export = true
 exporter.fhir_stu3.export = false
 exporter.fhir_dstu2.export = false
 exporter.fhir.use_shr_extensions = false
 exporter.fhir.use_us_core_ig = true
 exporter.fhir.transaction_bundle = true
-exporter.fhir.bulk_data = false
+exporter.fhir.bulk_data = true
 exporter.groups.fhir.export = false
 exporter.hospital.fhir.export = true
 exporter.hospital.fhir_stu3.export = false
@@ -68,9 +68,9 @@ exporter.bfd.clia_labs_count = 10
 exporter.bfd.cutoff_date=20140529

 exporter.cdw.export = false
-exporter.text.export = false
-exporter.text.per_encounter_export = false
-exporter.clinical_note.export = false
+exporter.text.export = true
+exporter.text.per_encounter_export = true
+exporter.clinical_note.export = true

**Comment by IoTFier** [comment]
- [ ] Build failed for Synthea.Properties Changed => exporter.fhir.bulk_data = true

100 warnings

> Task :test

org.mitre.synthea.export.CCDAExporterTest > testCCDAExport FAILED
    java.lang.AssertionError: Validation of exported CCDA failed: index=42, size=42 expected:<0> but was:<1>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:645)
        at org.mitre.synthea.export.CCDAExporterTest.testCCDAExport(CCDAExporterTest.java:71)

531 tests completed, 1 failed, 6 skipped

> Task :test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///D:/Dev/GitHub/Synthea/synthea/build/reports/tests/test/index.html

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 4m 34s
15 actionable tasks: 10 executed, 5 up-to-date
------------------------------
[failed_exports-09.08.22_set1.zip](https://github.com/synthetichealth/synthea/files/9528382/failed_exports-09.08.22_set1.zip)

**Comment by IoTFier** [comment]
Build failed for Synthea.Properties Changed => exporter.ccda.export = true

100 warnings

> Task :test

org.mitre.synthea.export.CCDAExporterTest > testCCDAExport FAILED
    java.lang.AssertionError: Validation of exported CCDA failed: Illegal operation (self.realmCode->isEmpty()) expected:<0> but was:<1>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:645)
        at org.mitre.synthea.export.CCDAExporterTest.testCCDAExport(CCDAExporterTest.java:71)

AppTest > testApp FAILED
    java.lang.AssertionError
        at org.junit.Assert.fail(Assert.java:86)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at org.junit.Assert.assertTrue(Assert.java:52)
        at AppTest.testApp(AppTest.java:53)

org.mitre.synthea.export.JSONExporterTest > export FAILED
    java.lang.AssertionError: Validation of exported JSON bundle failed: Module Dementia Module does not have an entered time in exported module history
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at org.mitre.synthea.export.JSONExporterTest.export(JSONExporterTest.java:60)

531 tests completed, 3 failed, 6 skipped

> Task :test FAILED

FAILURE: Build failed with an exception.

BUILD FAILED in 4m 31s
15 actionable tasks: 10 executed, 5 up-to-date

[failed_exports - 09.08.2022 Set 2.zip](https://github.com/synthetichealth/synthea/files/9528467/failed_exports.-.09.08.2022.Set.2.zip)

**Comment by IoTFier** [comment]
Hello @jawalonoski , @eedrummer  - 
I have done 6 separate builds this morning. All findings are attached in the messages above with failed exports folder as applicable -
1. First with undoing all my changes where build was successful
2. Next 5 builds were done one at a time for changing following parameters in synthea.properties file
- [ ] exporter.fhir.bulk_data = true | **_Build Failed; Failed Exports folder attached_** 
- [ ] exporter.ccda.export = true | **_Build Failed; Failed Exports folder attached_** 
- [ ] exporter.text.export = true  | **_Build Failed; Failed Exports folder attached_**  
- [ ] exporter.text.per_encounter_export = true | **_Build Succeeded_** 
- [ ] exporter.clinical_note.export = true | **_Build Failed; Failed Exports folder attached_**

It is certainly a big issue around CCDA. I have provided you with all input that I can from these build failures.
Please let me know if a screenshare session would be helpful to you.

Thoughts?

**Comment by jawalonoski** [comment]
I have successfully built, tested, and executed the code on Windows and Mac with your properties changes a half dozen times, and only once encountered an error with C-CDA unit tests.

Based on the history of issues, and that a vast majority of stack traces center around C-CDA tests [there was one recent exception in another thread with JSON export], I am extremely confident that none of the properties is the issue (i.e., they are red herrings) and it entirely has to do with runtime anomalies of the C-CDA library we are using, evidenced by the fact that without fail, every single exported C-CDA document validates correctly.

At this point, you should be able to actually _run_ Synthea without problems (I have also tested this repeatedly on Windows and Mac with your properties settings) -- in the meantime, just ignore any C-CDA _unit test failures_ (in fact,  you don't even have to run the unit tests for the most part unless you are making developmental changes).

**Comment by IoTFier** [comment]
@jawalonoski - with my apologies in advance - I am not understanding 
1. why the build works on your computers and not mine when I have consistently experienced build failures on most of the time with successful build very few times? What is it I am missing on my computer?
2. How do I run Synthea without problems when I build is failing?
3. How to ignore / disable C-CDA unit tests or failures - especially when build fails?

Your guidance will be highly appreciated.
Thank you.

**Comment by jawalonoski** [comment]
> 1. why the build works on your computers and not mine when I have consistently experienced build failures on most of the time with successful build very few times? 

To be precise, the **compile** works on your computer (otherwise the unit tests would not run). Only the **unit tests** fail. The unit tests are not required to run Synthea.

> What is it I am missing on my computer?

Nothing. The C-CDA validation library is inconsistent, and appears to be a memory issue. Memory issues have a habit of behaving differently on different computers.

> 2. How do I run Synthea without problems when I build is failing?

```
./gradlew clean compileJava
./run_synthea
```

> 3. How to ignore / disable C-CDA unit tests or failures - especially when build fails?

Ignore?

```
./gradlew clean compileJava
./run_synthea
```

Disable? 

Easiest way is to delete the `src/test/java/org/mitre/synthea/export/CCDAExporterTest.java` file.

**Comment by IoTFier** [comment]
Hi Jason. sorry for responding late. Your input helps.
Hopefully the C-CDA library team fixes that memory leak issue soon.

I still request you to keep this issue open, in case I find any build errors outside of CCDA test case.
Hope that's ok.

Have a good weekend!

**Comment by jawalonoski** [comment]
We believe this issue has been solved with https://github.com/synthetichealth/synthea/pull/1186.
